src/mse/main_media_source_interface.ts

```diff
@@ -539,7 +537,90 @@ export class MainSourceBufferInterface implements ISourceBufferInterface {
     }
     log.debug("SBI: pushing segment", this.type);
-    sourceBuffer.appendBuffer(data);
+
+    if (canRelyOnMseTimestampOffset) {
+      this._appendBufferWithTimestampOffsetCapabilities(data, timestampOffset);
+    } else {
+      this._appendBufferWithoutTimestampOffsetCapabilities(data, timestampOffset);
+    }
+  }
+
+  /**
+   * Append media or initialization segment (represented by `data`) to the given
+   * SourceBuffer (represented by `sourceBuffer`), with the given
+   * `timestampOffset` in seconds for devices where the MSE `timestampOffset`
+   * attribute **CAN** be used.
+   *
+   * @param {BufferSource} data
+   * @param {number|undefined} timestampOffset
+   */
+  private _appendBufferWithTimestampOffsetCapabilities(
+    data: BufferSource,
+    timestampOffset: number | undefined,
+  ): void {
+    if (
+      timestampOffset !== undefined &&
+      this._sourceBuffer.timestampOffset !== timestampOffset
+    ) {
+      const newTimestampOffset = timestampOffset;
+      log.debug(
+        "SBI: updating timestampOffset",
+        this._sourceBuffer.timestampOffset,
+        newTimestampOffset,
+      );
+      this._sourceBuffer.timestampOffset = newTimestampOffset;
+    }
+
+    this._sourceBuffer.appendBuffer(data);
+  }
+
+  /**
+   * Append media or initialization segment (represented by `data`) to the given
+   * SourceBuffer (represented by `sourceBuffer`), with the given
+   * `timestampOffset` in seconds for devices where the MSE `timestampOffset`
+   * attribute **CANNOT** be used.
+   *
+   * @param {BufferSource} data
+   * @param {number|undefined} timestampOffset
+   */
+  private _appendBufferWithoutTimestampOffsetCapabilities(
+    data: BufferSource,
+    timestampOffset: number | undefined,
+  ): void {
+    const dataU8 = toUint8Array(data);
+
+    // TODO also for WebM
+    const isMp4InitSegment = hasInitSegment(dataU8);
+    if (isMp4InitSegment) {
+      const mdhdTimeScale = getMDHDTimescale(dataU8);
+      this._initTimescale = mdhdTimeScale;
+    }
+
+    if (timestampOffset !== 0 && timestampOffset !== undefined) {
+      const initTimescale = this._initTimescale;
+      if (initTimescale === undefined) {
+        log.warn("Compat: not able to mutate decode time due to unknown timescale");
+        return this._appendBufferWithTimestampOffsetCapabilities(data, timestampOffset);
+      }
+      const oldTrackFragmentDecodeTime = getTrackFragmentDecodeTime(dataU8);
+      if (oldTrackFragmentDecodeTime === undefined) {
+        log.warn("Compat: not able to mutate decode time due to not found tfdt");
+        return this._appendBufferWithTimestampOffsetCapabilities(data, timestampOffset);
+      }
+
+      const timescaledOffset = initTimescale * timestampOffset;
+      const newTrackFragmentDecodeTime = oldTrackFragmentDecodeTime + timescaledOffset;
+      log.debug(
+        "Compat: Trying to update decode time",
+        oldTrackFragmentDecodeTime,
+        newTrackFragmentDecodeTime,
+      );
```

Comment:
We can add a symbol in the log so we have a hint while debugging on what's the old/new value:
ex: `Compat: Trying to update decode time 100 >>> 200` ?

Reply:
Good idea, will do